rust/osm2lanes/src/transform/tags_to_lanes/separator.rs

```diff
@@ -0,0 +1,150 @@
+use super::*;
+
+#[allow(clippy::needless_collect)]
+pub fn insert_separators(lanes: Lanes) -> LanesResult {
+    let Lanes { lanes, warnings } = lanes;
+    let left_edge = lane_to_edge_separator(lanes.first().unwrap());
+    let right_edge = lane_to_edge_separator(lanes.first().unwrap());
+    let separators: Vec<Option<Lane>> = lanes
+        .windows(2)
+        .map(|window| lanes_to_separator(window.try_into().unwrap(), &lanes))
+        .collect();
+    // I promise this is good code, but it might need a little explanation.
```

Comment:
This is reasonable to understand. An alternate idea, take it or leave it: build up an output list as you go:

```
let mut output_lanes = Vec::new();
for pair in lanes.windows(2) {
  if output_lanes.is_empty() {
    output_lanes.push(lane_to_edge_separator(pair[0]));
  }
  output_lanes.push(pair[0]);
  if let Some(separator) = lanes_to_separator(pair) {
    output_lanes.push(separator);
  }  
}
output_lanes.push(lane_to_edge_separator(output_lanes.last().unwrap());
```

Maybe even simpler to do the left and right edge stuff outside the loop entirely

Reply:
not convinced by yours either. I am too proud of my iterator solution to delete it now :P